bytecomp/translprim.ml

```diff
@@ -341,6 +344,7 @@ let primitives_table =
     "%bswap_native", Primitive ((Pbbswap(Pnativeint)), 1);
     "%int_as_pointer", Primitive (Pint_as_pointer, 1);
     "%opaque", Primitive (Popaque, 1);
+    "%sys_argv", External prim_sys_argv;
```

Comment:
I'm not fond of the way you describe an arity-0 primitive by rewriting into an arity-1 external call, that is in turn handled specially by the compiler later on. In the meantime, typing is wrong, and I'm not fully confident that this will never leak out.

Would it be reasonable to have a new `Primitive (Argv, 0)` primitive that is elaborated into `Lprim(Pccall prim_sys_get_argv, [Lconst (Const_pointer 0)], loc)`. (Note the `get_argv` name to highlight that it is a function, not a value.)

Reply:
I initially implemented this as I think you're suggesting, by extending Lambda.primitive with a new primitive. The awkward consequence was that the rewrite of `%sys_argv` to `caml_sys_argv ()` had to happen in multiple places, for the bytecode, native and Flambda native compilers. Putting the rewrite in `simplif.ml` gets bytecode and non-Flambda native, putting it in `cmmgen.ml` gets native and Flambda native, but I don't see anywhere that gets all three without duplication. The cleanest thing seems to be putting it in `translprim.ml`, so that Lambda code never refers to `%sys_argv` and the backends don't need to handle it.

I don't think the typing is wrong: `%sys_argv` is a 0-ary builtin primitive, and `caml_sys_argv` is a 1-ary C call. The external `%sys_argv`, of type `string array`, gets rewritten to the expression `caml_sys_argv ()`, also of type `string array`.

The name is `caml_sys_argv` mainly for bootstrapping: during development, there needed to be several bootstraps supporting both the old and the new primitive, so it couldn't have the same name as the existing `caml_sys_get_argv`.

Reply:
The expansion you make of `External prim_sys_argv` in simplif makes it well-typed again, but I would claim that, morally, there is a typing rule for Lambda code where `External primitive` gets a type from the arity of `prim_sys_argv`, and the Lambda code between the primitive translation and the simplification is not well-typed according to this rule.

I think that I would slightly prefer duplication over the current state, but my gut feeling is that it should be possible to avoid the External temporary state without having to duplicate. Maybe my suggesting to have it a `Primitive` is wrong, and it should be an additional constructor to `bytecomp/translprim.prim`?